src/compiler/checker.ts

```diff
@@ -3954,7 +3954,8 @@ namespace ts {
                     if (strictNullChecks && declaration.flags & NodeFlags.Ambient && isParameterDeclaration(declaration)) {
                         parentType = getNonNullableType(parentType);
                     }
-                    const propType = getTypeOfPropertyOfType(parentType, text);
+                    const parentTypeNEUndefined = getTypeWithFacts(parentType, TypeFacts.NEUndefined);

```

Comment:
1. It looks like undefined gets removed. whether or not there is an initialiser. I don't think this is correct.
2. Shouldn't there be an equivalent line for arrays? You can do things like `([first = 1, second = 2]) => first + second`, right?

Reply:
Here undefined should not be removed, but getTypeOfPropertyOfType should "see" through the levels of Union types (with undefined) and get the correct type of the property. Maybe the best approach here is to fix getTypeOfPropertyOfType instead of patching it here.

Reply:
The problem originates in ```getPropertyOfUnionOrIntersectionType()```, where the return type is ```undefined``` because the property is partial (only present in one of the constituents of the union).
This makes sense for some cases, but not in this (buggy) one. Probably the most "elegant" solution is to pass some sort of flag down the line to indicate if we want partials resolved or not.